src/components/AdvertisingSlot.js

```diff
@@ -1,39 +1,69 @@
-import React, { Component } from 'react';
-import connectToAdServer from './utils/connectToAdServer';
 import PropTypes from 'prop-types';
+import React, { useEffect, useRef, useContext } from 'react';
+import AdvertisingContext from '../AdvertisingContext';
+import calculateRootMargin from './utils/calculateRootMargin';
+import isLazyLoading from './utils/isLazyLoading';
+import getLazyLoadConfig from './utils/getLazyLoadConfig';
 
-class AdvertisingSlot extends Component {
-  componentDidMount() {
-    const { activate, id, customEventHandlers } = this.props;
-    activate(id, customEventHandlers);
-  }
-
-  componentDidUpdate(prevProps) {
-    const { activate, id, customEventHandlers } = this.props;
-    if (prevProps.activate !== activate) {
-      activate(id, customEventHandlers);
+function AdvertisingSlot({
+  id,
+  style,
+  className,
+  children,
+  customEventHandlers,
+}) {
+  const observerRef = useRef(null);
+  const containerDivRef = useRef();
+  const { activate, config } = useContext(AdvertisingContext);
+  const lazyLoadConfig = getLazyLoadConfig(config, id);
+  const isLazyLoadEnabled = isLazyLoading(lazyLoadConfig);
+  useEffect(() => {
+    if (!config || !isLazyLoadEnabled) {
+      return () => {};
     }
-  }
-
-  render() {
-    const { id, style, className, children } = this.props;
-    return (
-      <div id={id} style={style} className={className} children={children} />
+    const rootMargin = calculateRootMargin(lazyLoadConfig);
+    observerRef.current = new IntersectionObserver(
+      ([{ isIntersecting }]) => {
+        if (isIntersecting) {
+          activate(id, customEventHandlers);
+          observerRef.current.unobserve(containerDivRef.current);
+        }
+      },
+      { rootMargin }
     );
-  }
+    observerRef.current.observe(containerDivRef.current);
+    return () => {
+      observerRef.current.unobserve(containerDivRef.current);
+    };
+  }, [activate, config]);
+
+  useEffect(() => {
+    if (!config || isLazyLoadEnabled) {
+      return;
+    }
+    activate(id, customEventHandlers);
+  }, [activate, config]);
```

Comment:
This effect does not include the behaviour of the lifecycle method "componentDidMount".
Should we play on the safe side and include it?
`useEffect(() => { activate(id, customEventHandlers); }, []);`

Reply:
To my knowledge, `useEffect` executes when the component mounts as soon as you provide an array as second argument, regardless of the array containing elements or not. Otherwise the slot activation would not work at all with this code.

I actually made me a code sandbox to get to the bottom of this question some time ago:

https://codesandbox.io/s/use-effect-2et6c?file=/src/Schubidu.js

If you look at the log of the sandbox, you can see that the log statements from all three *useEffect* calls in component *Schubidi* execute when the sandbox is rendered the first time.